storage/index.go

```diff
@@ -21,61 +21,272 @@
 package storage
 
 import (
+	"bytes"
 	"errors"
+	"sync"
+	"time"
 
 	"github.com/m3db/m3db/storage/index"
+	"github.com/m3db/m3db/storage/namespace"
+	"github.com/m3db/m3ninx/doc"
+	"github.com/m3db/m3ninx/index/segment"
+	"github.com/m3db/m3ninx/index/segment/mem"
 	"github.com/m3db/m3x/context"
 	"github.com/m3db/m3x/ident"
+
+	"github.com/uber-go/tally"
 )
 
 var (
-	errIndexingNotImplemented = errors.New("indexing is not implemented")
+	errDbIndexAlreadyClosed                      = errors.New("database index has already been closed")
+	errDbIndexUnableToWriteClosed                = errors.New("unable to write to database index, already closed")
+	errDbIndexUnableToQueryClosed                = errors.New("unable to query database index, already closed")
+	errDbIndexUnableToIndexWithReservedFieldName = errors.New("unable to index document due to usage of reserved fieldname")
+)
+
+type nsIndexState byte
+
+const (
+	nsIndexStateOpen nsIndexState = iota
+	nsIndexStateClosed
 )
 
-type dbIndex struct {
-	opts Options
+type nsIndexBlock struct {
+	segment    mem.Segment
+	expiryTime time.Time
 }
 
-func newDatabaseIndex(o Options) (databaseIndex, error) {
-	return &dbIndex{
-		opts: o,
-	}, nil
+type nsIndex struct {
+	sync.RWMutex
+	insertMode index.InsertMode
+	state      nsIndexState
+	active     nsIndexBlock
+
+	insertQueue namespaceIndexInsertQueue
+	metrics     nsIndexMetrics
+	opts        index.Options
+	nsID        ident.ID
 }
 
-func (i *dbIndex) Write(
-	ctx context.Context,
-	namespace ident.ID,
+// nolint: deadcode
+func newNamespaceIndex(
+	md namespace.Metadata,
+	newIndexQueueFn newNamespaceIndexInsertQueueFn,
+	opts index.Options,
+) (namespaceIndex, error) {
+	if err := opts.Validate(); err != nil {
+		return nil, err
+	}
+
+	now := opts.ClockOptions().NowFn()()
+	segmentID := segment.ID(now.UnixNano())
+	seg, err := mem.New(segmentID, opts.MemSegmentOptions())
+	if err != nil {
+		return nil, err
+	}
+
+	expiryTime := now.Add(md.Options().RetentionOptions().RetentionPeriod())
+	idx := &nsIndex{
+		insertMode: opts.InsertMode(),
+		active: nsIndexBlock{
+			segment:    seg,
+			expiryTime: expiryTime, // FOLLOWUP(prateek): compute based on block rotation
+		},
+		metrics: newNamespaceIndexMetrics(opts.InstrumentOptions().MetricsScope()),
+		opts:    opts,
+		nsID:    md.ID(),
+	}
+
+	queue := newIndexQueueFn(idx.writeBatch, opts.ClockOptions().NowFn(),
+		opts.InstrumentOptions().MetricsScope())
+	if err := queue.Start(); err != nil {
+		return nil, err
+	}
+	idx.insertQueue = queue
+
+	return idx, nil
+}
+
+// NB(prateek): including the call chains leading to this point:
+//
+// - For new entry (previously unseen in the shard):
+//     shard.WriteTagged()
+//       => shardInsertQueue.Insert()
+//       => shard.writeBatch()
+//       => index.Write()
+//       => indexQueue.Insert()
```

Comment:
Is there a reason we don't just `index.writeBatch` from `shard.writeBatch`?  Seems like a second level of queuing that may not be necessary?

Reply:
talked offline, going to keep the two level indexing going to allow for batch'd writes to the index.